src/Abstractions/Entities/TaskEntity.cs

```diff
@@ -71,176 +69,51 @@ public interface ITaskEntity
-    {
-        Check.NotNull(operation);
-        if (!this.TryDispatchMethod(operation, out object? result, out Type returnType))
-        {
-            throw new NotSupportedException($"No suitable method found for entity operation '{operation}'.");
-        }
-
-        if (typeof(Task).IsAssignableFrom(returnType))
-        {
-            // Task or Task<T>
-            return new(AsGeneric((Task)result!, returnType)); // we assume a declared Task return type is never null.
-        }
-
-        if (returnType == typeof(ValueTask))
-        {
-            // ValueTask
-            return AsGeneric((ValueTask)result!); // we assume a declared ValueTask return type is never null.
-        }
-
-        if (returnType.IsGenericType && returnType.GetGenericTypeDefinition() == typeof(ValueTask<>))
-        {
-            // ValueTask<T>
-            return AsGeneric(result!, returnType); // No inheritance, have to do purely via reflection.
-        }
-
-        return new(result);
-    }
-
-    static bool TryGetInput(ParameterInfo parameter, TaskEntityOperation operation, out object? input)
-    {
-        if (!operation.HasInput)
-        {
-            if (parameter.HasDefaultValue)
-            {
-                input = parameter.DefaultValue;
-                return true;
-            }
-
-            input = null;
-            return false;
-        }
-
-        input = operation.GetInput(parameter.ParameterType);
-        return true;
-    }
-
-    static async Task<object?> AsGeneric(Task task, Type declared)
-    {
-        await task;
-        if (declared.IsGenericType && declared.GetGenericTypeDefinition() == typeof(Task<>))
-        {
-            return declared.GetProperty("Result", BindingFlags.Public | BindingFlags.Instance).GetValue(task);
-        }
-
-        return null;
-    }
-
-    static ValueTask<object?> AsGeneric(ValueTask t)
-    {
-        static async Task<object?> Await(ValueTask t)
-        {
-            await t;
-            return null;
-        }
+    /// <summary>
+    /// Gets a value indicating whether dispatching operations to <see cref="State"/> is allowed. State dispatch will
+    /// only be attempted if entity-level dispatch does not succeed. Default is <c>true</c>.
+    /// </summary>
+    protected bool AllowStateDispatch => true;
 
-        if (t.IsCompletedSuccessfully)
-        {
-            return default;
-        }
+    /// <summary>
+    /// Gets or sets the state for this entity.
+    /// </summary>
+    /// <remarks>
+    /// This will be hydrated as part of <see cref="RunAsync(TaskEntityOperation)"/>. The contents of this property
+    /// will be persisted to <see cref="TaskEntityContext.SetState(object?)"/> when the operation completes. Deleting
+    /// entity state can be accomplished by setting this to default(<typeparamref name="TState"/>).
+    /// </remarks>
+    protected TState State { get; set; } = default!; // leave null-checks to end implementation.
```

Comment:
Yes the `!` operator is for suppressing nullability warnings. You can generalize it as appropriate for use when static analysis does not work for your scenario. In our case, it doesn't work, we don't know the deserialization logic of `TState`, we don't know if they want to throw on `null` or not. 

Although, after some thought I think `TState?` is appropriate here. Because we will initialize to `default(TState)`. I just want to verify it works as intended with structs.

Reply:
I am undecided if `TState` or `TState?` is better here. The common scenario is for `TState` to be non-null for most operation dispatches (since we initialize via `Activator.CreateInstance` now). However, there are a couple scenarios where `null` is okay when interacting with this property:

1. Deleting state would be done via `this.State = default!;` (which is `null` for reference types).
2. Overriding `InitializeState` and setting to `null` is also allowed - it is then up to users to decide how `null` state behaves.

So, which is better? `TState?` to show that it _can_ be null, and then users need to use `!.` everywhere. OR `TState` to avoid the `!` because it is typically _not_ null.

Reply:
How about `Optional<TState>` and then make TState non-nullable?

Since the backend really does make a distinction between "not present" (meaning there is no record for this entity in storage) and "default value" it may be just be safer to maintain that distinction throughout the API. Otherwise weird things can happen, especially for types where the default value is not null.

Reply:
I can see that `Optional<TState>` would be readonly so we can't just use that one directly. But I think we can copy its semantics, so we could have
1. a boolean property `HasState` which determines whether the entity has state or not, and
2. a non-null `TState State` whose content is meaningful only if `HasState` is true

Reply:
Per our offline discussion, I have switched back to `TState` from `TState?`. We will not use an optional type and have an experience of only deleting state when it is `null` (and **not** default of a struct).

Reply:
> We will add a future implicit delete operation which will delete state for reference or value type.

Can we add this operation now?

Reply:
I have the implicit delete code ready for another PR.